go/vt/etcdtopo/client_test.go

```diff
@@ -185,8 +254,49 @@ func (c *fakeClient) SetCluster(machines []string) bool {
 
 func (c *fakeClient) Watch(prefix string, waitIndex uint64, recursive bool,
 	receiver chan *etcd.Response, stop chan bool) (*etcd.Response, error) {
-	c.Lock()
-	defer c.Unlock()
 
-	return &etcd.Response{}, nil
+	// We need a buffered forwarderfor 2 reasons:
+	// - in the select loop below, we only write to receiver if
+	// stop has not been closed. Otherwise we introduce race
+	// conditions.
+	// - we are waiting on forwarder and taking the node mutex.
+	// fakeNode.notify write to forwarder, and also takes the node
+	// mutex. Both can deadlock each-other. By buffering the
+	// channel, we make sure one notify() call can finish and not
+	// deadlock, which is as many as we do in the unit tests.
+	forwarder := make(chan *etcd.Response, 1)
+
+	// add the watch under the lock
+	c.Lock()
+	if recursive {
+		panic("not implemented")
+	}
+	c.createParentDirs(prefix)
+	n, ok := c.nodes[prefix]
+	if !ok {
+		n = newFakeNode(nil)
+		c.nodes[prefix] = n
+	}
+	c.Unlock()
+
+	n.mu.Lock()
+	index := n.watchIndex
+	n.watchIndex++
+	n.watches[index] = forwarder
+	n.mu.Unlock()
+
+	// and wait until we stop, each action will write to forwarder, send
+	// these along. The order in select is critical: we only want to
```

Comment:
If the order matters, you need to use two selects. Each select chooses pseudo-randomly with uniform distribution if more than one case is ready to proceed.